File: gds/src/projection/abstract/ElementProjection.ts
import { DefaultValue } from "@/api";
import { Aggregation } from "@/core/Aggregation";
import { PropertyMapping } from "./PropertyMapping";
import { PropertyMappings } from "../primitive/PropertyMappings";
import { PropertyMappingsBuilder } from "../primitive/PropertyMappings";

/**
 * Base class for projections of graph elements (nodes or relationships).
 */
export abstract class ElementProjection {
  /**
   * Wildcard symbol to project all properties.
   */
  public static readonly PROJECT_ALL = "*";

  /**
   * Key used for properties in configuration objects.
   */
  public static readonly PROPERTIES_KEY = "properties";

  /**
   * Returns the property mappings for this projection.
   */
  public properties(): PropertyMappings {
    return PropertyMappings.of();
  }

  /**
   * Creates a new projection with additional property mappings.
   *
   * @param mappings The mappings to add
   */
  public abstract withAdditionalPropertyMappings(
    mappings: PropertyMappings
  ): ElementProjection;

  /**
   * Checks if this projection includes all properties.
   */
  public abstract projectAll(): boolean;

  /**
   * Converts this projection to a plain object.
   */
  public toObject(): Record<string, any> {
    const value: Record<string, any> = {};
    this.writeToObject(value);
    value[ElementProjection.PROPERTIES_KEY] = this.properties().toObject(
      this.includeAggregation()
    );
    return value;
  }

  /**
   * Writes this projection's specific properties to the given object.
   *
   * @param value The object to write to
   */
  protected abstract writeToObject(value: Record<string, any>): void;

  /**
   * Checks if aggregation should be included in serialized objects.
   */
  protected abstract includeAggregation(): boolean;

  /**
   * Creates an ElementProjection from a configuration object.
   *
   * @param config The configuration object
   * @param constructor Function that creates a projection from property mappings
   */
  public static create<T extends ElementProjection>(
    config: Record<string, any>,
    constructor: (propertyMappings: PropertyMappings) => T
  ): T {
    const properties = config[ElementProjection.PROPERTIES_KEY] ?? {};
    const propertyMappings = PropertyMappings.fromObject(properties);
    return constructor(propertyMappings);
  }

  /**
   * Gets a non-empty string from a configuration object.
   *
   * @param config The configuration object
   * @param key The key to lookup
   * @returns The string value
   * @throws Error if the value is not a non-empty string
   */
  protected static nonEmptyString(
    config: Record<string, any>,
    key: string
  ): string {
    const value = config[key];
    if (typeof value !== "string" || value.length === 0) {
      throw new Error(`'${value}' is not a valid value for the key '${key}'`);
    }
    return value;
  }
}

/**
 * Interface for classes that support inline property addition.
 */
export interface InlineProperties<Self extends InlineProperties<Self>> {
  /**
   * Adds a property mapping.
   *
   * @param mapping The mapping to add
   */
  addProperty(mapping: PropertyMapping): Self;

  /**
   * Adds a property with the given key, neo property key, and default value.
   *
   * @param propertyKey The property key
   * @param neoPropertyKey The neo property key
   * @param defaultValue The default value
   */
  addProperty(
    propertyKey: string,
    neoPropertyKey: string,
    defaultValue: DefaultValue
  ): Self;

  /**
   * Adds a property with the given key, neo property key, default value, and aggregation.
   *
   * @param propertyKey The property key
   * @param neoPropertyKey The neo property key
   * @param defaultValue The default value
   * @param aggregation The aggregation
   */
  addProperty(
    propertyKey: string,
    neoPropertyKey: string,
    defaultValue: DefaultValue,
    aggregation: Aggregation
  ): Self;

  /**
   * Adds multiple property mappings.
   *
   * @param properties The mappings to add
   */
  addProperties(...properties: PropertyMapping[]): Self;

  /**
   * Adds all property mappings from an iterable.
   *
   * @param properties The mappings to add
   */
  addAllProperties(properties: Iterable<PropertyMapping>): Self;

  /**
   * Finalizes the property building process.
   */
  buildProperties(): void;

  /**
   * Gets the inline properties builder.
   */
  inlineBuilder(): InlinePropertiesBuilder;
}

/**
 * Builder for inline properties.
 */
export class InlinePropertiesBuilder {
  // Change this from PropertyMappings.Builder to PropertyMappingsBuilder
  private _propertiesBuilder: PropertyMappingsBuilder | null = null;

  /**
   * Creates a new InlinePropertiesBuilder.
   *
   * @param getProperties Function to get current properties
   * @param setProperties Function to set new properties
   */
  constructor(
    private readonly getProperties: () => PropertyMappings | null,
    private readonly setProperties: (
      properties: PropertyMappings | null
    ) => void
  ) {}

  /**
   * Builds the property mappings.
   */
  public build(): void {
    if (this._propertiesBuilder !== null) {
      if (this.getProperties() !== null) {
        throw new Error(
          "Cannot have both, a complete mapping from `properties` " +
            "and other properties from `addProperty`. If you want to " +
            "combine those, make sure to call `properties` first and " +
            "then use `addProperty` and never set a new `properties` " +
            "again."
        );
      }
      this.setProperties(this._propertiesBuilder.build());
    }
  }

  /**
   * Gets the property mappings builder.
   */
  // Change the return type from PropertyMappings.Builder to PropertyMappingsBuilder
  public propertiesBuilder(): PropertyMappingsBuilder {
    if (this._propertiesBuilder === null) {
      this._propertiesBuilder = PropertyMappings.builder();
      const properties = this.getProperties();
      if (properties !== null) {
        this._propertiesBuilder.from(properties);
        this.setProperties(null);
      }
    }
    return this._propertiesBuilder;
  }
}

File: gds/src/projection/traits/element_projection.rs
use super::property_mapping::PropertyMapping;
use crate::projection::{PropertyMappings, PropertyMappingsBuilder};
use std::collections::HashMap;

/// Wildcard symbol to project all properties.
pub const PROJECT_ALL: &str = "*";

/// Key used for properties in configuration objects.
pub const PROPERTIES_KEY: &str = "properties";

/// Base trait for projections of graph elements (nodes or relationships).
///
/// Defines the interface for element projections with property mappings.
pub trait ElementProjection: Send + Sync {
    /// Returns the property mappings for this projection.
    fn properties(&self) -> &PropertyMappings;

    /// Creates a new projection with additional property mappings.
    fn with_additional_property_mappings(
        &self,
        mappings: PropertyMappings,
    ) -> Box<dyn ElementProjection>;

    /// Checks if this projection includes all properties.
    fn project_all(&self) -> bool;

    /// Checks if aggregation should be included in serialized output.
    fn include_aggregation(&self) -> bool;

    /// Converts this projection to a configuration map.
    fn to_config(&self) -> HashMap<String, serde_json::Value>;
}

/// Trait for types that support inline property addition.
///
/// Mirrors the Java GDS inline projection API while delegating to the shared PropertyMappings builder.
pub trait InlineProperties: Sized {
    /// Gets the inline properties builder.
    fn inline_builder(&mut self) -> &mut InlinePropertiesBuilder;

    /// Adds a property mapping.
    fn add_property_mapping(mut self, mapping: PropertyMapping) -> Self {
        self.inline_builder()
            .properties_builder()
            .add_mapping_ref(mapping);
        self
    }

    /// Adds a simple property by key.
    fn add_property(mut self, property_key: impl Into<String>) -> Result<Self, String> {
        let mapping = PropertyMapping::of(property_key)?;
        self.inline_builder()
            .properties_builder()
            .add_mapping_ref(mapping);
        Ok(self)
    }

    /// Adds a property with source name.
    fn add_property_with_source(
        mut self,
        property_key: impl Into<String>,
        neo_property_key: impl Into<String>,
    ) -> Result<Self, String> {
        let mapping = PropertyMapping::with_source(property_key, neo_property_key)?;
        self.inline_builder()
            .properties_builder()
            .add_mapping_ref(mapping);
        Ok(self)
    }

    /// Adds multiple property mappings.
    fn add_all_properties(mut self, mappings: Vec<PropertyMapping>) -> Self {
        for mapping in mappings {
            self.inline_builder()
                .properties_builder()
                .add_mapping_ref(mapping);
        }
        self
    }

    /// Finalizes the property building process.
    fn build_properties(&mut self) {
        self.inline_builder().build();
    }
}

/// Helper for building properties inline during projection construction.
#[derive(Debug)]
pub struct InlinePropertiesBuilder {
    properties_builder: Option<PropertyMappingsBuilder>,
    properties: Option<PropertyMappings>,
}

impl InlinePropertiesBuilder {
    /// Creates a new InlinePropertiesBuilder.
    pub fn new() -> Self {
        InlinePropertiesBuilder {
            properties_builder: None,
            properties: None,
        }
    }

    /// Creates a builder with existing properties.
    pub fn with_properties(properties: PropertyMappings) -> Self {
        InlinePropertiesBuilder {
            properties_builder: None,
            properties: Some(properties),
        }
    }

    /// Gets or creates the properties builder.
    pub fn properties_builder(&mut self) -> &mut PropertyMappingsBuilder {
        if self.properties_builder.is_none() {
            let builder = if let Some(ref properties) = self.properties {
                PropertyMappingsBuilder::new().from(properties)
            } else {
                PropertyMappingsBuilder::new()
            };
            self.properties = None;
            self.properties_builder = Some(builder);
        }
        self.properties_builder.as_mut().unwrap()
    }

    /// Builds the final PropertyMappings.
    pub fn build(&mut self) -> Option<PropertyMappings> {
        if let Some(builder) = self.properties_builder.take() {
            if self.properties.is_some() {
                panic!(
                    "Cannot have both complete mapping from `properties` \
                     and other properties from `add_property`"
                );
            }
            Some(builder.build())
        } else {
            self.properties.take()
        }
    }

    /// Gets the current properties (if not being built).
    pub fn get_properties(&self) -> Option<&PropertyMappings> {
        self.properties.as_ref()
    }

    /// Sets the properties directly.
    pub fn set_properties(&mut self, properties: PropertyMappings) {
        self.properties = Some(properties);
        self.properties_builder = None;
    }
}

impl Default for InlinePropertiesBuilder {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_property_mappings_empty() {
        let mappings = PropertyMappings::empty();
        assert!(mappings.is_empty());
        assert_eq!(mappings.size(), 0);
    }

    #[test]
    fn test_property_mappings_builder() {
        let mappings = PropertyMappings::builder()
            .add_mapping(PropertyMapping::of("age").unwrap())
            .add_mapping(PropertyMapping::with_source("score", "user_score").unwrap())
            .build();

        assert_eq!(mappings.size(), 2);
        let keys = mappings.property_keys();
        assert!(keys.contains("age"));
        assert!(keys.contains("score"));
    }

    #[test]
    fn test_property_mappings_merge() {
        let mappings1 = PropertyMappings::builder()
            .add_mapping(PropertyMapping::of("age").unwrap())
            .build();

        let mappings2 = PropertyMappings::builder()
            .add_mapping(PropertyMapping::of("score").unwrap())
            .build();

        let merged = mappings1.merge_with(&mappings2);
        assert_eq!(merged.size(), 2);
        let keys = merged.property_keys();
        assert!(keys.contains("age"));
        assert!(keys.contains("score"));
    }

    #[test]
    fn test_inline_properties_builder() {
        let mut builder = InlinePropertiesBuilder::new();
        builder
            .properties_builder()
            .add_mapping_ref(PropertyMapping::of("age").unwrap());

        let properties = builder.build().unwrap();
        assert_eq!(properties.size(), 1);
        assert!(properties.property_keys().contains("age"));
    }
}
